tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc

```diff
@@ -46,27 +48,51 @@ class ConvertSelectV2 : public OpConverterBase<ConvertSelectV2> {
 
     const auto& params = *this->params_;
     const auto& inputs = params.inputs;
-    const auto& cond_input = inputs.at(0);
+    const auto& i_cond = inputs.at(0);
     const auto& node = params.node_def;
     TF_RETURN_IF_ERROR(
-        check_type(cond_input.TrtDType(), nvinfer1::DataType::kBOOL, node));
+        check_type(i_cond.TrtDType(), nvinfer1::DataType::kBOOL, node));
 
-    if (cond_input.is_weights()) {
+    if (i_cond.is_weights()) {
+      // According to
+      // https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#constant-layer
+      // Boolean weights are not supported in TRT version 8.4.
       return errors::InvalidArgument(bool_weight_error_msg(node));
     }
 
-    const auto type_then = inputs[1].TrtDType();
-    const auto type_else = inputs[2].TrtDType();
+    const auto& i_then = inputs.at(1);
+    const auto& i_else = inputs.at(2);
+    const auto type_then = i_then.TrtDType();
+    const auto type_else = i_else.TrtDType();
     if (type_then != type_else && (type_then == nvinfer1::DataType::kINT32 ||
                                    type_else == nvinfer1::DataType::kINT32)) {
       // Both or none of (type_then, type_else) should be equal to kINT32.
       return errors::InvalidArgument(
           then_else_dtypes_error_msg(type_then, type_else, node));
     }
 
+    bool cond_is_vector = false;
+    const auto& shape_cond = i_cond.GetTrtDims();
+    if (layer_name_ == "select") {
+      const auto& shape_then = i_then.GetTrtDims();
+      const auto& shape_else = i_else.GetTrtDims();
+      TF_RETURN_IF_ERROR(compare_shapes(shape_then, shape_else));
+      TF_RETURN_IF_ERROR(
+          compare_shapes(shape_cond, shape_then, &cond_is_vector));
+    }
+
+    nvinfer1::Dims cond_dims(shape_cond);
+    if (cond_is_vector) {
+      cond_dims.nbDims = i_then.GetTrtDims().nbDims;
+      const std::vector<int> ones(cond_dims.d[0], 1);
+      std::copy(ones.begin(), ones.end(), cond_dims.d + 1);
+    }
+
+    const TRT_TensorOrWeights new_cond(nvinfer1::DataType::kBOOL, cond_dims,
+                                       i_cond.batch_size());
     nvinfer1::Dims broadcasted_dims[3];
     for (int i = 1; i < 3; i++) {
-      TF_RETURN_IF_ERROR(GetTrtBroadcastShape(cond_input, inputs.at(i), true,
+      TF_RETURN_IF_ERROR(GetTrtBroadcastShape(new_cond, inputs.at(i), true,
                                               false, broadcasted_dims,
                                               broadcasted_dims + i));
     }
```

Comment:
Commenting here, but it relates to line 102: AFAIK PrepareTensorForShape would fail for tensors with dynamic shapes, I wonder why does it pass the tests? Consider using [ApplyBroadcast](https://github.com/tensorflow/tensorflow/blob/1ee420e303bdef0ac6c8e9a2b352627001782f18/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L421)

Reply:
I will try to use [ApplyBroadcast](https://github.com/tensorflow/tensorflow/blob/1ee420e303bdef0ac6c8e9a2b352627001782f18/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L421), but my first few attempts were unsuccessful.